### PR TITLE
Add ExpectType to eslint-plugin

### PR DESCRIPTION
**Pull request by jakebailey**
This moves the expect rule from tslint to eslint. By default in the editor, it will run ExpectType on the builtin version of TypeScript. But when run via dtslint, it will run all TS versions (and also convert diagnostics into lint errors).

This leaves npm-naming as the only remaining tslint rule.

TODO:

- [x] Multiple versions
- [x] Only run one version in editor, run all otherwise
- [x] Don't return compiler errors in editor
- [x] Delete old rule
- [ ] ~Delete tslint???~
- [x] Perf test now that I am running all TS versions, or go back to the hacky way
- [x] Never call diagnostics from the editor-based setup

### CODE REVIEW

File: packages/eslint-plugin/src/rules/expect.ts [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+      messageId: "programContents",
+      data: { fileName, versionName },
+      loc: zeroSourceLocation,
+    });
+    return;
+  }
+
+  const checker = program.getTypeChecker();
+  // Don't care about emit errors.
+  const diagnostics = ts.getPreEmitDiagnostics(program, sourceFile);
```
Comment (jakebailey) [inline packages/eslint-plugin/src/rules/expect.ts:227]: I'll need to drop this call if we're using the `node_modules` copy of typescript; the editor will already show the errors and as of https://github.com/typescript-eslint/typescript-eslint/pull/7997 this is no longer safe.

File: packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+==== types/expect/index.d.ts ====
+
+    export const foo = 1234;
+    
+    export const aUnion: string | number | undefined;
+    
+    export function complicatedUnion<T extends string | number>(x: T, y: T): {
+        prop1: "a" | "b" | "c";
+        prop2: readonly (string | number)[];
```
Comment (sandersn) [inline packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint:15]:
```suggestion
        prop2: ReadonlyArray<string | number>;
```

**Review by sandersn — APPROVED** [review]
Just a couple of small comments/questions.

File: packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
[the remark this answers is not in the document]
Comment (jakebailey) [reply packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint:16]: The baselines run all rules; the naming of the folder has nothing to do with what tests affect it unless we add eslintrcs to each directory.

File: packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint (hunk 2 above)
Comment (jakebailey) [reply packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint:15]: This is intentional to test that I can use one syntax or the other in the rule that normalizes the form.

File: packages/dtslint/src/lint.ts [hunk 4]
```diff
@@ -64,63 +56,46 @@ export async function lint(
   }
   const result = linter.getResult();
```
[the remark this answers is not in the document]
Comment (jakebailey) [reply packages/dtslint/src/lint.ts:57]: Ah, good point.

File: packages/dtslint/src/lint.ts [hunk 5]
```diff
@@ -64,63 +56,46 @@ export async function lint(
...
+            "@definitelytyped/expect": ["error", { versionsToTest }],
+          },
+        },
+      ],
+    },
+  };
+
+  if (expectOnly) {
+    // Disable the regular config, instead load only the plugins and use just the rule above.
+    // TODO(jakebailey): share this with eslint-plugin
```
[the remark this answers is not in the document]
Comment (jakebailey) [reply packages/dtslint/src/lint.ts:81]: Yeah, augmenting it was clearer to me as it'd be a good bit of duplication.

File: packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandersn) [reply packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint:16]: Right, but I expected you'd write the test to only have expect failures. Upon re-reading the sorting code, I see some manipulation of ReadonlyArray, so maybe that's it? But the error still comes from array-type.

[the remark this answers is not in the document]
Comment (jakebailey) [reply packages/eslint-plugin/test/__file_snapshots__/types/expect/index.d.ts.lint:16]: Right; this is mainly to test that the `ReadonlyArray` normalization I reimplemented still works.

File: packages/eslint-plugin/test/__file_snapshots__/types/expect/expect-tests.ts.lint [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+!!! @definitelytyped/expect: Cannot match a node to this assertion. If this is a multiline function call, ensure the assertion is on the line above.
+    
+    
+    // $ExpectType NotMatched
+    
+    ~
+!!! @definitelytyped/expect: TypeScript expected type to be:
+!!!                        :   NotMatched
+!!!                        : got:
+!!!                        :   any
```
Comment (jakebailey) [inline packages/eslint-plugin/test/__file_snapshots__/types/expect/expect-tests.ts.lint:86]: It's not super clear to me whether or not this is a new bug or not; this should complain that it's unmatched. I'll have to test with the old code, but I did not change the logic so I assume that this is preexisting.

**Comment by jakebailey** [comment]
> Not necessary to address here, but does it really make sense to keep ExpectRule responsible for reporting program diagnostics?

It's only the expect rule that loops over every TS version, so it's the only one that can report these errors. I have disabled that reporting when in the editor though (since tsserver is there), but if we want to do it outside of the plugin we're going to have to move the looping logic over to dtslint and then duplicate the programs too (or, do some sort of program sharing).

File: packages/eslint-plugin/test/__file_snapshots__/types/expect/expect-tests.ts.lint (hunk 6 above)
Comment (jakebailey) [reply packages/eslint-plugin/test/__file_snapshots__/types/expect/expect-tests.ts.lint]: Ah, it was finding `EndOfFileToken`. Fixed.

**Pull request merged**